platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/DefaultTestEventGenerator.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.tasks.testing.TestEventGenerator;
+import org.gradle.api.tasks.testing.TestFailure;
+import org.gradle.api.tasks.testing.TestOutputEvent;
+import org.gradle.api.tasks.testing.TestResult;
+
+import javax.annotation.Nullable;
+import java.time.Instant;
+
+@NonNullApi
+public class DefaultTestEventGenerator implements TestEventGenerator {
+    @NonNullApi
+    protected enum State {
+        CREATED, STARTED, COMPLETED, CLOSED
```

Comment:
💭 I think (in some future iteration) there needs to be unit tests verifying this state maintenance.  

I also kind of wonder why we need track this state?  If the caller messes up and calls completed twice, and we have 2 completed events, or they don't call started, what breaks?

Reply:
We pass it directly to tooling API and build ops, so I'm not sure what would break, but it could cause UI glitches or crashes.

Reply:
Another question for the Tooling API experts, I suppose.